packages/host/tests/helpers/adapter.ts

```diff
@@ -0,0 +1,299 @@
+        throw new TokenExpiredError(`JWT token expired at ${expiration}`);
+      }
+      return claims;
+    }
+    throw new JsonWebTokenError(`unable to verify JWT: ${token}`);
+  }
+
+  get lastModified() {
+    return this.#lastModified;
+  }
+
+  // this is to aid debugging since privates are actually not visible in the debugger
+  get files() {
+    return this.#files;
+  }
+
+  async *readdir(
+    path: string,
+  ): AsyncGenerator<{ name: string; path: string; kind: Kind }, void> {
+    let dir =
+      path === '' ? this.#files : this.#traverse(path.split('/'), 'directory');
+    for (let [name, content] of Object.entries((dir as Dir).contents)) {
+      yield {
+        name,
+        path: path === '' ? name : `${path}/${name}`,
+        kind: content.kind,
+      };
+    }
+  }
+
+  async exists(path: string): Promise<boolean> {
+    let maybeFilename = path.split('/').pop()!;
+    try {
+      // a quirk of our test file system's traverse is that it creates
+      // directories as it goes--so do our best to determine if we are checking for
+      // a file that exists (because of this behavior directories always exist)
+      await this.#traverse(
+        path.split('/'),
+        maybeFilename.includes('.') ? 'file' : 'directory',
+      );
+      return true;
+    } catch (err: any) {
+      if (err.name === 'NotFoundError') {
+        return false;
+      }
+      if (err.name === 'TypeMismatchError') {
+        try {
+          await this.#traverse(path.split('/'), 'file');
+          return true;
+        } catch (err: any) {
+          if (err.name === 'NotFoundError') {
+            return false;
+          }
+          throw err;
+        }
+      }
+      throw err;
+    }
+  }
+
+  async openFile(
+    path: LocalPath,
+    loader: Loader,
+  ): Promise<FileRef | undefined> {
+    let content;
+    try {
+      content = this.#traverse(path.split('/'), 'file');
+    } catch (err: any) {
+      if (['TypeMismatchError', 'NotFoundError'].includes(err.name)) {
+        return undefined;
+      }
+      throw err;
+    }
+    if (content.kind === 'directory') {
+      return undefined;
+    }
+
+    let value = content.content;
+
+    let fileRefContent = '';
+
+    if (path.endsWith('.json')) {
+      let cardApi = await loader.import<CardAPI>(`${baseRealm.url}card-api`);
+      if (cardApi.isCard(value)) {
+        value.id = `${this.#paths.url}${path.replace(/\.json$/, '')}`;
+        cardApi.setCardAsSavedForTest(value);
+        let doc = cardApi.serializeCard(value);
+        fileRefContent = JSON.stringify(doc);
+      } else {
+        fileRefContent =
+          typeof value === 'string' ? value : JSON.stringify(value);
+      }
+    } else if (path.endsWith('.gts')) {
+      if (typeof value === 'string') {
+        fileRefContent = value;
+      } else {
+        let moduleURLString = `${this.#paths.url}${path.replace(/\.gts$/, '')}`;
+
+        loader.shimModule(moduleURLString, value as object);
```

Comment:
This shimming is needed for serializing cards to make content for the adapter. Without shimming the serializing would fail because the module would be missing from the identity map.

Reply:
Another option one could consider here would be to load the module using the loader to get the module into the identity map, but we would get into an infinite recursion because `openFile` would be called and we'd end up in this exact same spot